File: CodeGenA32/isel_tester.cc
// (c) Robert Muth - see LICENSE for more info

#include "Base/serialize.h"
#include "CodeGenA32/isel_gen.h"
#include "CodeGenA32/regs.h"
#include "CpuA32/symbolic.h"
#include "Util/parse.h"

#include <iostream>

namespace {

using namespace cwerg;
using namespace cwerg::base;

const char* OpTypeString(Handle op) {
  switch (op.kind()) {
    default:
      return "_";
    case RefKind::CONST:
      return EnumToString(ConstKind(Const(op)));
    case RefKind::REG:
      return EnumToString(RegKind(Reg(op)));
  }
}

void HandleIns(Ins ins) {
  std::cout << "INS: ";
  InsRenderToAsm(ins, &std::cout);
  std::cout << "  [";
  const char* sep = "";
  for (unsigned i = 0; i < InsOpcode(ins).num_operands; ++i) {
    std::cout << sep << OpTypeString(InsOperand(ins, i));
    sep = " ";
  }
  std::cout << "]\n";
  if (InsOPC(ins) == OPC::NOP1 || InsOPC(ins) == OPC::RET) {
    std::cout << "    SPECIAL\n";
    return;
  }

  const uint8_t mismatches =
      code_gen_a32::FindtImmediateMismatchesInBestMatchPattern(ins);
  if (mismatches == code_gen_a32::MATCH_IMPOSSIBLE) {
    std::cout << "    MATCH_IMPOSSIBLE\n";
  } else if (mismatches > 0) {
    std::cout << "    mismatches: " << std::hex << (unsigned)mismatches
              << std::dec << "\n";
  } else {
    const code_gen_a32::Pattern* pat = code_gen_a32::FindMatchingPattern(ins);
    ASSERT(pat != nullptr, "");
    std::cout << "PAT: reg:[";
    sep = "";
    for (unsigned i = 0; i < InsOpcode(ins).num_operands; ++i) {
      std::cout << sep << EnumToString(pat->reg_constraints[i]);
      sep = " ";
    }
    std::cout << "]  imm:[";
    sep = "";
    for (unsigned i = 0; i < InsOpcode(ins).num_operands; ++i) {
      std::cout << sep << code_gen_a32::EnumToString(pat->imm_constraints[i]);
      sep = " ";
    }
    std::cout << "]\n";

    code_gen_a32::EmitContext ctx;
    for (unsigned i = 0; i < pat->length; ++i) {
      const code_gen_a32::InsTmpl& tmpl = pat->start[i];
      a32::Ins a32ins = code_gen_a32::MakeInsFromTmpl(tmpl, ins, ctx);
      char buffer[128];
      a32::RenderInsSystematic(a32ins, buffer);
      std::cout << "    " << buffer << "\n";
    }
  }
}

void Process(std::istream* input) {
  code_gen_a32::InitCodeGenA32();
  std::vector<char> data = SlurpDataFromStream(input);

  Unit unit = UnitParseFromAsm("unit", {data.data(), data.size()},
                               code_gen_a32::GetAllRegs());
  // UnitRenderToAsm(unit, &std::cout);
  for (Fun fun : UnitFunIter(unit)) {
    for (Bbl bbl : FunBblIter(fun)) {
      for (Ins ins : BblInsIter(bbl)) {
        std::cout << "\n";
        HandleIns(ins);
      }
    }
  }
}

}  // namespace

int main(int argc, const char* argv[]) {
  InitStripes(1);
  Process(&std::cin);
  return 0;
}

File: CodeGenA32/isel_tester.py
#!/usr/bin/python3
"""Testing helper for table driven code selection"""

from typing import Any
import sys

from Base import serialize
from Base import ir

from BE.CpuA32 import opcode_tab as a32
from BE.CpuA32 import symbolic

from CodeGenA32 import isel_tab
from CodeGenA32 import regs


def OpToStr(op: Any) -> str:
    if isinstance(op, (isel_tab.PARAM, a32.PRED, a32.REG, a32.SHIFT)):
        return op.name
    assert isinstance(op, int)
    return str(op)


def OpTypeStr(op: Any) -> str:
    if isinstance(op, ir.Reg):
        return op.kind.name
    elif isinstance(op, ir.Const):
        return op.kind.name
    else:
        return "_"


def HandleIns(ins: ir.Ins, ctx: regs.EmitContext):
    print("INS: " + serialize.InsRenderToAsm(
        ins).strip() + f"  [{' '.join(OpTypeStr(o) for o in ins.operands)}]")
    if ins.opcode in isel_tab.OPCODES_REQUIRING_SPECIAL_HANDLING:
        print(f"    SPECIAL")
        return
    mismatches = isel_tab.FindtImmediateMismatchesInBestMatchPattern(ins, False)
    if mismatches == isel_tab.MATCH_IMPOSSIBLE:
        print(f"    MATCH_IMPOSSIBLE")
    elif mismatches != 0:
        pattern = isel_tab.FindMatchingPattern(ins)
        assert pattern is None
        print(f"    mismatches: {mismatches:x}")
    else:
        pattern = isel_tab.FindMatchingPattern(ins)
        print(
            f"PAT: reg:[{' '.join(a.name for a in pattern.type_curbs)}]  "
            f"imm:[{' '.join(a.name for a in pattern.imm_curbs)}]")
        for tmpl in pattern.emit:
            armins = tmpl.MakeInsFromTmpl(ins, ctx)
            name, ops = symbolic.InsSymbolize(armins)
            print(f"    {name} {' '.join(ops)}")


def Translate(fin):
    unit = serialize.UnitParseFromAsm(fin, cpu_regs=regs.CPU_REGS_MAP)
    for fun in unit.funs:
        ctx = regs.EmitContext(0xfc0, 0xfc0, 0xffff0000, 0xffff0000, 66)
        if "gpr_scratch" in fun.name:
            ctx.scratch_cpu_reg = regs.GPR_REGS[6]
        fun.FinalizeStackSlots()
        for bbl in fun.bbls:
            for ins in bbl.inss:
                print()
                HandleIns(ins, ctx)


if __name__ == "__main__":
    Translate(sys.stdin)
